types/algoliasearch.d.ts

```diff
@@ -0,0 +1,148 @@
+// Note: Below, we will be importing all algoliasearch v3,4,5 types.
+// The goal is being able to export the algoliasearch-helper types using
+// the version of the client installed by the developer.
+
+// @ts-ignore
+import type algoliasearch from 'algoliasearch/lite';
+// @ts-ignore
+import type * as AlgoliaSearchLite from 'algoliasearch/lite';
+// @ts-ignore
+import type * as AlgoliaSearch from 'algoliasearch';
+// @ts-ignore
+import type * as ClientSearch from '@algolia/client-search';
+
+// turns any to unknown, so it can be used as a conditional
+type AnyToUnknown<T> = (any extends T ? true : false) extends true
+  ? unknown
+  : T;
+
+type DummySearchClientV4 = {
+  transporter: unknown;
+};
+
+type DummySearchClient = {
+  search: unknown;
+};
+
+// @ts-ignore
+type ClientV3_4 = ReturnType<typeof algoliasearch>;
+
+type ClientLiteV5 = AnyToUnknown<
+  // @ts-ignore
+  ReturnType<typeof AlgoliaSearchLite.liteClient>
+>;
+type ClientFullV5 = AnyToUnknown<
+  // @ts-ignore
+  ReturnType<typeof AlgoliaSearch.algoliasearch>
+>;
+type ClientSearchV5 = AnyToUnknown<
+  // @ts-ignore
+  ReturnType<typeof ClientSearch.searchClient>
+>;
+type ClientV5 = ClientLiteV5 extends DummySearchClient
+  ? ClientLiteV5
+  : ClientSearchV5 extends DummySearchClient
+  ? ClientSearchV5
+  : ClientFullV5 extends DummySearchClient
+  ? ClientFullV5
+  : unknown;
+
+type PickForClient<
+  T extends {
+    v3: unknown;
+    v4: unknown;
+    v5: unknown;
+  }
+> = ClientV5 extends DummySearchClient
+  ? T['v5']
+  : // @ts-ignore
+  ClientV3_4 extends DummySearchClientV4
+  ? T['v4']
+  : T['v3'];
+
+type DefaultSearchClient = PickForClient<{
+  v3: ClientV3_4;
+  v4: ClientV3_4;
+  v5: ClientV5;
+}>;
+
+export type SearchOptions = PickForClient<{
+  // @ts-ignore
+  v3: AlgoliaSearch.QueryParameters;
+  // @ts-ignore
+  v4: ClientSearch.SearchOptions;
+  v5: NonNullable<
+    // @ts-ignore
+    AlgoliaSearch.LegacySearchMethodProps[number]['params']
+  >;
+}>;
+
+export type SearchResponse<T> = PickForClient<{
+  // @ts-ignore
+  v3: AlgoliaSearch.Response<T> & {
+    appliedRelevancyStrictness?: number;
+    nbSortedHits?: number;
+    renderingContent?: {
+      facetOrdering?: {
+        facets?: {
+          order?: string[];
+        };
+        values?: {
+          [facet: string]: {
+            order?: string[];
+            sortRemainingBy?: 'count' | 'alpha' | 'hidden';
+          };
+        };
+      };
+    };
+  };
+  // @ts-ignore
+  v4: ClientSearch.SearchResponse<T>;
+  // @ts-ignore
+  v5: AlgoliaSearch.SearchResponse; // TODO: should be generic
```

Comment:
debating with @shortcuts whether SearchResponse already can have this from the alpha, as you can't give a generic to something that doesn't have it, or detect when something is or isn't generic

Reply:
if not possible, we can ship like this and later add the generic when it will be generic, but that means there's a moment where v5 is supported but the response isn't generic in eg. hits

Reply:
Why isn't the v5 `SearchResponse` generic for now?

Reply:
It's generated code so we did not (yet) focused on implementing it in some places (`search`/`browse` for example), but will do before stable!

Reply:
This means we'll need to provide a default value in the helper when it becomes generic so it doesn't break, which is fine. Do you see any possible issues @Haroenv?

Reply:
The issue I see is that we will have to say "if you're using 3.9.0 of the helper, it's supported, if you're using 3.10.0 (or whatever version has the final) it's generic". It's not that big of a deal as you can just say to update of course